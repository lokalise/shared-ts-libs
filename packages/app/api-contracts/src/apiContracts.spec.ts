import { describe, expect, it } from 'vitest'
import { z } from 'zod'
import {
  buildDeleteRoute,
  buildGetRoute,
  buildPayloadRoute,
  mapRouteToPath,
} from './apiContracts.ts'

const BODY_SCHEMA = z.object({})
const PATH_PARAMS_SCHEMA = z.object({
  userId: z.string(),
})
const PATH_PARAMS_MULTI_SCHEMA = z.object({
  userId: z.string(),
  orgId: z.string(),
})

describe('apiContracts', () => {
  describe('buildPayloadRoute', () => {
    it('sets default change route values', () => {
      const contract = buildPayloadRoute({
        successResponseBodySchema: BODY_SCHEMA,
        requestBodySchema: BODY_SCHEMA,
        method: 'post',
        description: 'some description',
        responseSchemasByStatusCode: {
          200: BODY_SCHEMA,
          400: z.object({ message: z.string() }),
        },
        pathResolver: () => '/',
      })

      expect(contract).toMatchSnapshot()
    })
  })

  describe('buildGetRoute', () => {
    it('sets default get route values', () => {
      const contract = buildGetRoute({
        successResponseBodySchema: BODY_SCHEMA,
        pathResolver: () => '/',
        responseSchemasByStatusCode: {
          '200': BODY_SCHEMA,
          '400': z.object({ message: z.string() }),
        },
        description: 'some description',
        metadata: { hello: 'world' }, // it is a free form record
      })

      expect(contract).toMatchSnapshot()
    })

    it('resolves path params', () => {
      const contract = buildGetRoute({
        successResponseBodySchema: BODY_SCHEMA,
        requestPathParamsSchema: PATH_PARAMS_SCHEMA,
        pathResolver: (pathParams) => `/users/${pathParams.userId}`,
      })

      expect(contract).toMatchSnapshot()
    })
  })

  describe('buildDeleteRoute', () => {
    it('sets default delete route values', () => {
      const contract = buildDeleteRoute({
        successResponseBodySchema: BODY_SCHEMA,
        pathResolver: () => '/',
        description: 'some description',
      })

      expect(contract).toMatchSnapshot()
    })
  })

  describe('mapRouteToPath', () => {
    it('returns path without params', () => {
      const contract = buildGetRoute({
        successResponseBodySchema: BODY_SCHEMA,
        pathResolver: () => '/',
      })

      const path = mapRouteToPath(contract)
      expect(path).toEqual('/')
    })

    it('returns path with one param', () => {
      const contract = buildGetRoute({
        successResponseBodySchema: BODY_SCHEMA,
        requestPathParamsSchema: PATH_PARAMS_SCHEMA,
        pathResolver: (pathParams) => `/users/${pathParams.userId}`,
      })

      const path = mapRouteToPath(contract)
      expect(path).toEqual('/users/:userId')
    })

    it('returns path with multiple params', () => {
      const contract = buildGetRoute({
        successResponseBodySchema: BODY_SCHEMA,
        requestPathParamsSchema: PATH_PARAMS_MULTI_SCHEMA,
        pathResolver: (pathParams) => `/orgs/${pathParams.orgId}/users/${pathParams.userId}`,
      })

      const path = mapRouteToPath(contract)
      expect(path).toEqual('/orgs/:orgId/users/:userId')
    })
  })
})
