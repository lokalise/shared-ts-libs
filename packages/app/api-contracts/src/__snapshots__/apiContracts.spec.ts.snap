// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`apiContracts > buildDeleteRoute > sets default delete route values 1`] = `
{
  "description": "some description",
  "isEmptyResponseExpected": true,
  "isNonJSONResponseExpected": false,
  "metadata": undefined,
  "method": "delete",
  "pathResolver": [Function],
  "requestHeaderSchema": undefined,
  "requestPathParamsSchema": undefined,
  "requestQuerySchema": undefined,
  "responseSchemasByStatusCode": undefined,
  "successResponseBodySchema": ZodObject {
    "_cached": null,
    "_def": {
      "catchall": ZodNever {
        "_def": {
          "typeName": "ZodNever",
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "pipe": [Function],
        "promise": [Function],
        "readonly": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "shape": [Function],
      "typeName": "ZodObject",
      "unknownKeys": "strip",
    },
    "and": [Function],
    "array": [Function],
    "augment": [Function],
    "brand": [Function],
    "catch": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nonstrict": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "pipe": [Function],
    "promise": [Function],
    "readonly": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "~standard": {
      "validate": [Function],
      "vendor": "zod",
      "version": 1,
    },
  },
  "summary": undefined,
  "tags": undefined,
}
`;

exports[`apiContracts > buildGetRoute > resolves path params 1`] = `
{
  "description": undefined,
  "isEmptyResponseExpected": false,
  "isNonJSONResponseExpected": false,
  "metadata": undefined,
  "method": "get",
  "pathResolver": [Function],
  "requestHeaderSchema": undefined,
  "requestPathParamsSchema": ZodObject {
    "_cached": null,
    "_def": {
      "catchall": ZodNever {
        "_def": {
          "typeName": "ZodNever",
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "pipe": [Function],
        "promise": [Function],
        "readonly": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "shape": [Function],
      "typeName": "ZodObject",
      "unknownKeys": "strip",
    },
    "and": [Function],
    "array": [Function],
    "augment": [Function],
    "brand": [Function],
    "catch": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nonstrict": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "pipe": [Function],
    "promise": [Function],
    "readonly": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "~standard": {
      "validate": [Function],
      "vendor": "zod",
      "version": 1,
    },
  },
  "requestQuerySchema": undefined,
  "responseSchemasByStatusCode": undefined,
  "successResponseBodySchema": ZodObject {
    "_cached": null,
    "_def": {
      "catchall": ZodNever {
        "_def": {
          "typeName": "ZodNever",
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "pipe": [Function],
        "promise": [Function],
        "readonly": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "shape": [Function],
      "typeName": "ZodObject",
      "unknownKeys": "strip",
    },
    "and": [Function],
    "array": [Function],
    "augment": [Function],
    "brand": [Function],
    "catch": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nonstrict": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "pipe": [Function],
    "promise": [Function],
    "readonly": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "~standard": {
      "validate": [Function],
      "vendor": "zod",
      "version": 1,
    },
  },
  "summary": undefined,
  "tags": undefined,
}
`;

exports[`apiContracts > buildGetRoute > sets default get route values 1`] = `
{
  "description": "some description",
  "isEmptyResponseExpected": false,
  "isNonJSONResponseExpected": false,
  "metadata": {
    "hello": "world",
  },
  "method": "get",
  "pathResolver": [Function],
  "requestHeaderSchema": undefined,
  "requestPathParamsSchema": undefined,
  "requestQuerySchema": undefined,
  "responseSchemasByStatusCode": {
    "200": ZodObject {
      "_cached": null,
      "_def": {
        "catchall": ZodNever {
          "_def": {
            "typeName": "ZodNever",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "shape": [Function],
        "typeName": "ZodObject",
        "unknownKeys": "strip",
      },
      "and": [Function],
      "array": [Function],
      "augment": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nonstrict": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "400": ZodObject {
      "_cached": null,
      "_def": {
        "catchall": ZodNever {
          "_def": {
            "typeName": "ZodNever",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "shape": [Function],
        "typeName": "ZodObject",
        "unknownKeys": "strip",
      },
      "and": [Function],
      "array": [Function],
      "augment": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nonstrict": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
  },
  "successResponseBodySchema": ZodObject {
    "_cached": null,
    "_def": {
      "catchall": ZodNever {
        "_def": {
          "typeName": "ZodNever",
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "pipe": [Function],
        "promise": [Function],
        "readonly": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "shape": [Function],
      "typeName": "ZodObject",
      "unknownKeys": "strip",
    },
    "and": [Function],
    "array": [Function],
    "augment": [Function],
    "brand": [Function],
    "catch": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nonstrict": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "pipe": [Function],
    "promise": [Function],
    "readonly": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "~standard": {
      "validate": [Function],
      "vendor": "zod",
      "version": 1,
    },
  },
  "summary": undefined,
  "tags": undefined,
}
`;

exports[`apiContracts > buildPayloadRoute > sets default change route values 1`] = `
{
  "description": "some description",
  "isEmptyResponseExpected": false,
  "isNonJSONResponseExpected": false,
  "metadata": undefined,
  "method": "post",
  "pathResolver": [Function],
  "requestBodySchema": ZodObject {
    "_cached": null,
    "_def": {
      "catchall": ZodNever {
        "_def": {
          "typeName": "ZodNever",
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "pipe": [Function],
        "promise": [Function],
        "readonly": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "shape": [Function],
      "typeName": "ZodObject",
      "unknownKeys": "strip",
    },
    "and": [Function],
    "array": [Function],
    "augment": [Function],
    "brand": [Function],
    "catch": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nonstrict": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "pipe": [Function],
    "promise": [Function],
    "readonly": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "~standard": {
      "validate": [Function],
      "vendor": "zod",
      "version": 1,
    },
  },
  "requestHeaderSchema": undefined,
  "requestPathParamsSchema": undefined,
  "requestQuerySchema": undefined,
  "responseSchemasByStatusCode": {
    "200": ZodObject {
      "_cached": null,
      "_def": {
        "catchall": ZodNever {
          "_def": {
            "typeName": "ZodNever",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "shape": [Function],
        "typeName": "ZodObject",
        "unknownKeys": "strip",
      },
      "and": [Function],
      "array": [Function],
      "augment": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nonstrict": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "400": ZodObject {
      "_cached": null,
      "_def": {
        "catchall": ZodNever {
          "_def": {
            "typeName": "ZodNever",
          },
          "and": [Function],
          "array": [Function],
          "brand": [Function],
          "catch": [Function],
          "default": [Function],
          "describe": [Function],
          "isNullable": [Function],
          "isOptional": [Function],
          "nullable": [Function],
          "nullish": [Function],
          "optional": [Function],
          "or": [Function],
          "parse": [Function],
          "parseAsync": [Function],
          "pipe": [Function],
          "promise": [Function],
          "readonly": [Function],
          "refine": [Function],
          "refinement": [Function],
          "safeParse": [Function],
          "safeParseAsync": [Function],
          "spa": [Function],
          "superRefine": [Function],
          "transform": [Function],
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "shape": [Function],
        "typeName": "ZodObject",
        "unknownKeys": "strip",
      },
      "and": [Function],
      "array": [Function],
      "augment": [Function],
      "brand": [Function],
      "catch": [Function],
      "default": [Function],
      "describe": [Function],
      "isNullable": [Function],
      "isOptional": [Function],
      "nonstrict": [Function],
      "nullable": [Function],
      "nullish": [Function],
      "optional": [Function],
      "or": [Function],
      "parse": [Function],
      "parseAsync": [Function],
      "pipe": [Function],
      "promise": [Function],
      "readonly": [Function],
      "refine": [Function],
      "refinement": [Function],
      "safeParse": [Function],
      "safeParseAsync": [Function],
      "spa": [Function],
      "superRefine": [Function],
      "transform": [Function],
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
  },
  "successResponseBodySchema": ZodObject {
    "_cached": null,
    "_def": {
      "catchall": ZodNever {
        "_def": {
          "typeName": "ZodNever",
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "pipe": [Function],
        "promise": [Function],
        "readonly": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "shape": [Function],
      "typeName": "ZodObject",
      "unknownKeys": "strip",
    },
    "and": [Function],
    "array": [Function],
    "augment": [Function],
    "brand": [Function],
    "catch": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nonstrict": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "pipe": [Function],
    "promise": [Function],
    "readonly": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "~standard": {
      "validate": [Function],
      "vendor": "zod",
      "version": 1,
    },
  },
  "summary": undefined,
  "tags": undefined,
}
`;
